keps/sig-auth/0014-dynamic-audit-configuration.md

```diff
@@ -148,19 +155,32 @@ backend:
+As a cluster admin, I will easily be able to enable the internal auditing features of an existing cluster, and tweak the configurations as necessary. I want to prevent privilege escalation from being able to tamper with a root audit configuration.
 
 #### Story 2
 As a Kubernetes extension developer, I will be able to provide drop in extensions that utilize audit data.
 
 #### Story 3
 As a cluster admin, I will be able configure multiple audit-policies and webhook endpoints to provide independent auditing facilities.
 
+#### Story 4
+As a kubernetes developer, I will be able to quickly turn up the audit level on a certain area to debug my application.
+
 ### Implementation Details/Notes/Constraints
 
+#### Feature Gating
+Introduction of dynamic policy requires changes to the current audit pipeline. Care must be taken that these changes are properly gated and do not affect the stability of the current features as they progress to GA. A new decorated handler will be provisioned similar to the [existing handlers](https://github.com/kubernetes/apiserver/blob/master/pkg/endpoints/filters/audit.go#L41) called `withDynamicAudit`. Another conditional clause will be added where the handlers are [provisioned](https://github.com/kubernetes/apiserver/blob/master/pkg/server/config.go#L536) allowing for the proper feature gating.
+
+#### Filtering
+This addition will move policy enforcement from the main handler to the backends. From the `withDynamicAudit` handler, the full event will be generated and then passed to the backends. Each backend will copy the event and then be required to drop any pieces that do not conform to its policy. A new sink interface will be required for these changes called `FilteredSink`, this will largely follow suite with the existing sink but take a fully formed event and the authorizer attributes as its parameters. It will then utilize the `LevelAndStages` method in the policy [checker](https://github.com/kubernetes/apiserver/blob/master/pkg/audit/policy/checker.go) to enforce its policy on the event, and drop any unneeded sections. The new dynamic backend will implement the `FilteredSink` interface, and update its state based on a shared informer. For the existing backends to comply, a `FilteredBackend` plugin will be built that can wrap existing backends with the new `FilteredSink` interface.
+
+Implementing the [attribute interface](/Users/patrick/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/authorization/authorizer/interfaces.go) based on the Event struct was also explored. This would allow us to keep the existing `Sink` interfaces, however it would require parsing the request URI twice in the pipeline due to how that field is represented in the Event. This was determined to not be worth the cost.
```

Comment:
I'm not sure I understand this, but parsing the request URI twice is likely to be much cheaper than copying the audit event N times.